File: app-geekhome/src/main/java/eu/geekhome/rest/HolderService.java
package eu.geekhome.rest;

import org.jvnet.hk2.annotations.Service;

import javax.ws.rs.core.Application;

@Service
public class HolderService<T> {

    private T _instance;

    public T getInstance() {
        return _instance;
    }

    public HolderService(Application app, Class<T> type) throws Exception {
        for (Object singleton : app.getSingletons()) {
            if (type.isAssignableFrom(singleton.getClass())) {
                _instance = (T) singleton;
                break;
            }
        }
        if (_instance == null) {
            throw new Exception("There's no singleton of type " + type.getName() +" in the app singletons container");
        }
    }
}

File: app-geekhome/src/main/java/eu/geekhome/rest/HolderService.kt
package eu.geekhome.rest

import org.jvnet.hk2.annotations.Service
import java.lang.Exception
import javax.ws.rs.core.Application

@Service
open class HolderService<T: Any>(app: Application, type: Class<T>) {
    lateinit var instance: T

    init {
        for (singleton in app.singletons) {
            if (type.isAssignableFrom(singleton.javaClass)) {

                @Suppress("UNCHECKED_CAST")
                instance = singleton as T
                break
            }
        }

        @Suppress("SENSELESS_COMPARISON")
        if (instance == null) {
            throw Exception("There's no singleton of type " + type.name + " in the app singletons container")
        }
    }
}